### PR TITLE
CT: Max Memory Size Check

### PR DESCRIPTION
**Pull request by facuMH**
Thanks to the fuzzing experiments an unchecked overflow was found and fixed in LFVM memory expansion calculation function. 

This PRs add that fix in the analogous CT function and a corresponding test case that fails without this newly introduced check.

### CODE REVIEW

File: go/ct/st/memory_test.go [hunk 1]
```diff
@@ -139,15 +139,19 @@ func TestMemory_MemoryExpansionCosts(t *testing.T) {
...
+		"large size no overflow":            {NewU256(0), NewU256(math.MaxUint64), MaxGas, 0, math.MaxUint64},
+		"large offset no overflow":          {NewU256(math.MaxUint64 - 1), NewU256(1), MaxGas, math.MaxUint64 - 1, 1},
+		"large offset and size no overflow": {NewU256(math.MaxUint64 / 2), NewU256(math.MaxUint64 / 2), MaxGas, math.MaxUint64 / 2, math.MaxUint64 / 2},
 		"size overflow":                     {NewU256(0), NewU256(1, 0), tosca.Gas(math.MaxInt64), 0, 0},
 		"offset overflow":                   {NewU256(1, 0), NewU256(1), tosca.Gas(math.MaxInt64), 0, 0},
 		"large offset and size overflow":    {NewU256(math.MaxUint64/2 + 1), NewU256(math.MaxUint64/2 + 1), tosca.Gas(math.MaxInt64), math.MaxUint64/2 + 1, math.MaxUint64/2 + 1},
 		"zero size":                         {NewU256(0), NewU256(0), tosca.Gas(0), 0, 0},
 		"zero size offset":                  {NewU256(1024), NewU256(0), tosca.Gas(0), 1024, 0},
 		"zero size offset overflow":         {NewU256(1, 0), NewU256(0), tosca.Gas(0), 0, 0},
+		"max memory size allowed":           {NewU256(0), NewU256(0x1FFFFFFFE1), MaxGas, 0, 0x1FFFFFFFE1},
```
Comment (simonlechner) [inline go/ct/st/memory_test.go:151]: Maybe add a test case with the max memory size + 1 that should overflow.

**Review by simonlechner — DISMISSED** [review]
Looks good.

File: go/ct/st/memory_test.go (hunk 1 above)
Comment (HerbertJordan) [inline go/ct/st/memory_test.go:151]: To improve readability, I would use `maxMemoryExpansionSize` instead of `0x1FFFFFFFE0` and `maxMemoryExpansionSize+1` instead of `0x1FFFFFFFE1`

File: go/ct/st/memory_test.go [hunk 2]
```diff
@@ -139,15 +139,19 @@ func TestMemory_MemoryExpansionCosts(t *testing.T) {
...
+		"acceptable size":                   {NewU256(0), NewU256(0x1FFFFFFFE0), tosca.Gas(36028809887088637), 0, 0x1FFFFFFFE0},
+		"acceptable offset":                 {NewU256(0x1FFFFFFFE0), NewU256(1), tosca.Gas(1152921504606846976), 0x1FFFFFFFE0, 1},
```
Comment (HerbertJordan) [inline go/ct/st/memory_test.go:153]: I do not understand this line. Why would this be acceptable?

File: go/ct/st/memory.go [hunk 3]
```diff
@@ -83,9 +83,23 @@ func (m *Memory) Grow(offset, size uint64) {
 	}
 }
 
+const (
+	// Maximum memory size allowed
+	// This magic number comes from 'core/vm/gas_table.go' 'memoryGasCost' in geth
```
Comment (HerbertJordan) [inline go/ct/st/memory.go:88]: Please provide a cleaner definition for this value. Saying that it is a "magic number from some source" is not building up trust.

File: go/ct/st/memory.go [hunk 4]
```diff
@@ -111,13 +125,20 @@ func (m *Memory) ExpansionCosts(offset_u256, size_u256 U256) (memCost tosca.Gas,
 		memCost = 0
 		return
 	}
+
+	if newSize > maxMemoryExpansionSize {
+		memCost = MaxGas
```
Comment (HerbertJordan) [inline go/ct/st/memory.go:130]: In the other cases where we have saturated memory costs we return `math.MaxInt64`. Here, it returns `MaxGas`. This is not consistent. I think this here should be `math.MaxInt64` as well.

Note that `MaxGas` is not `tosca.Gas(math.MaxInt64)`.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/st/memory_test.go (hunk 1 above)
Comment (facuMH) [reply go/ct/st/memory_test.go:151]: the `maxMemoryExpansionSize` is `0x1FFFFFFFE0`, so this case is actually testing the "+1" case. I will use the variable to make it obvious.

File: go/ct/st/memory.go (hunk 4 above)
Comment (facuMH) [reply go/ct/st/memory.go:130]: I am aware of the difference, I thought using `MaxGas` was better since it is smaller than `MaxInt64` but this is intentional, as described in #293. I think we should keep it consistent, yes, but that given that there is an existing issue about it, I thought sticking with `MaxGas` was the safer path. Should we remove the issue and remove all uses of `MaxGas` or stick with it and change other uses of `MaxInt64` to `MaxGas` ?

**Review by HerbertJordan — APPROVED** [review]
I belief you when you are saying this change is fixing something, but I do not see it in the PR. Can you add in the description something that lists a test case or a command that was not working before and is working now, that was not introduced specifically to check the function changed in this PR?

**Review by simonlechner — APPROVED** [review]
